crates/datafusion_ext/src/metrics.rs

```diff
@@ -0,0 +1,252 @@
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.child.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(Self::new(children[0].clone())))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        let stream = self.child.execute(partition, context)?;
+        Ok(Box::pin(BoxedStreamAdapater::new(
+            stream,
+            partition,
+            &self.metrics,
+        )))
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.child.statistics()
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+}
+
+impl DisplayAs for DataSourceMetricsExecAdapter {
+    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "DataSourceMetricsExecAdapter")
+    }
+}
+
+struct BoxedStreamAdapater {
+    stream: SendableRecordBatchStream,
+    metrics: DataSourceMetrics,
+}
+
+impl BoxedStreamAdapater {
+    fn new(
+        stream: SendableRecordBatchStream,
+        partition: usize,
+        metrics: &ExecutionPlanMetricsSet,
+    ) -> Self {
+        Self {
+            stream,
+            metrics: DataSourceMetrics::new(partition, metrics),
+        }
+    }
+}
+
+impl Stream for BoxedStreamAdapater {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let poll = self.stream.poll_next_unpin(cx);
+        self.metrics.record_poll(poll)
+    }
+}
+
+impl RecordBatchStream for BoxedStreamAdapater {
+    fn schema(&self) -> SchemaRef {
+        self.stream.schema()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AggregatedMetrics {
+    /// Total time taken for a plan to execute.
+    pub elapsed_compute_ns: u64,
+    /// Total output rows for the plan.
+    pub output_rows: u64,
+    /// Total bytes processed.
+    pub bytes_processed: u64,
+}
+
+impl AggregatedMetrics {
+    /// Computes aggregated metrics from a plan.
+    ///
+    /// The plan should have already been executed to completion, otherwise
+    /// partial or incorrect results will be reported.
+    pub fn new_from_plan(plan: &dyn ExecutionPlan) -> Self {
+        let mut agg = AggregatedMetrics {
+            elapsed_compute_ns: 0,
+            output_rows: 0,
+            bytes_processed: 0,
+        };
+        agg.aggregate_recurse(plan);
+        agg
+    }
+
+    fn aggregate_recurse(&mut self, plan: &dyn ExecutionPlan) {
+        // TODO: What to do about output rows?
+
+        let metrics = match plan.metrics() {
+            Some(metrics) => metrics,
+            None => return,
+        };
+        self.elapsed_compute_ns += metrics.elapsed_compute().unwrap_or_default() as u64;
```

Comment:
> I don't have all of the delicacies around duration types and time keeping types, but I've always thought nanosecond resolution ends up collecting a lot of bytes of noise data on most platforms (and therefore compresses relatively poorly for no value).

I agree, I figured just put this in since it's being tracked by some execs already (as ns). I do have concerns around the accuracy though, like I'm seeing 1ns for some ops which I know take a bit longer than 1ns.

> I tend to think that durations should in theory be able to be negative to indicate before (also detecting overflows (in the non-debug case) is potentially less obvious, but perhaps I'm on the wrong side of idioms.

This sounds very confusing to me in the context of plan execution. When would a node execute in negative time?

> Speaking of being on the wrong side of the idiom I'm always bemused when people trac duration in integers that are hilariously too large: we can track 580+ year long query execution time, though given parallelism...

Datafusion defines this metric as `usize`,